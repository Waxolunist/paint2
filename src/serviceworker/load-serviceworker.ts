import {html, render} from 'lit-html';
import {Workbox} from 'workbox-window';
import {TemplateResult} from 'lit-element';
import {Snackbar} from '@material/mwc-snackbar';

declare global {
  interface HTMLElementEventMap {
    'MDCSnackbar:opened': CustomEvent<unknown>;
    'MDCSnackbar:closed': CustomEvent<{reason: string}>;
  }
}

const snackbarTemplate = (): TemplateResult => html`
  <mwc-snackbar
    labelText="A new version is available. Press REFRESH to get the new version."
    timeoutMs="-1"
    stacked
  >
    <mwc-button slot="action">REFRESH</mwc-button>
    <mwc-icon-button icon="close" slot="dismiss"></mwc-icon-button>
  </mwc-snackbar>
`;

export const renderSnackbar = async (
  renderRootId = 'snackbar'
): Promise<Snackbar | null> => {
  await Promise.all([
    import('@material/mwc-snackbar'),
    import('@material/mwc-button'),
    import('@material/mwc-icon-button'),
  ]);
  const renderRoot = document.getElementById(renderRootId) || document.body;
  render(snackbarTemplate(), renderRoot);
  const snackbar: Snackbar | null = renderRoot.querySelector('mwc-snackbar');
  snackbar?.addEventListener(
    'MDCSnackbar:closed',
    (event: CustomEvent<{reason: string}>) => {
      if (event.detail.reason === 'action') window.location.reload();
    }
  );
  return snackbar;
};

if ('serviceWorker' in navigator && process.env.NODE_ENV === 'production') {
  const wb = new Workbox('/sw.js', {
    scope: new URL(document.baseURI).pathname,
  });

  wb.addEventListener('installed', (event) => {
    if (event.isUpdate) {
      renderSnackbar().then((snackbar) => snackbar?.show());
    }
  });

  wb.register();
}
