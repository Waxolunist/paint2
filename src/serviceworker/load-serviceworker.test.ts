import {fixture, html} from '@open-wc/testing-helpers';
import {renderSnackbar} from './load-serviceworker';

describe('snackbar', () => {
  it('renders correctly', async () => {
    const placeholderelement = await fixture(html`<div id="snackbar"></div>`);
    await renderSnackbar('snackbar');
    expect(placeholderelement!.innerHTML).toMatchSnapshot();
  });

  it('open', async () => {
    await fixture(html`<div><div id="snackbar"></div></div>`);
    const snackbar = await renderSnackbar('snackbar');
    expect(snackbar?.open).toBeFalsy();
    snackbar?.show();
    expect(snackbar?.open).toBeTruthy();
  });
});
