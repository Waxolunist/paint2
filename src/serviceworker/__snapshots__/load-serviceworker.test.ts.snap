// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snackbar renders correctly 1`] = `
"<!---->
  <mwc-snackbar labeltext=\\"A new version is available. Press REFRESH to get the new version.\\" timeoutms=\\"-1\\" stacked=\\"\\">
    <mwc-button slot=\\"action\\">REFRESH</mwc-button>
    <mwc-icon-button icon=\\"close\\" slot=\\"dismiss\\"></mwc-icon-button>
  </mwc-snackbar>
"
`;
